modules/evm-accounts/src/lib.rs

```diff
@@ -0,0 +1,229 @@
+use sp_io::{crypto::secp256k1_ecdsa_recover, hashing::keccak_256};
+use sp_runtime::traits::UniqueSaturatedInto;
+use sp_std::vec::Vec;
+
+mod default_weight;
+mod mock;
+mod tests;
+
+pub trait WeightInfo {
+	fn claim_account() -> Weight;
+}
+
+/// Evm Address.
+pub type EvmAddress = sp_core::H160;
+
+// FIXME: substrate new version for https://github.com/paritytech/substrate/pull/7216
+#[derive(Encode, Decode, Clone)]
+pub struct EcdsaSignature(pub [u8; 65]);
+
+impl PartialEq for EcdsaSignature {
+	fn eq(&self, other: &Self) -> bool {
+		&self.0[..] == &other.0[..]
+	}
+}
+
+impl sp_std::fmt::Debug for EcdsaSignature {
+	fn fmt(&self, f: &mut sp_std::fmt::Formatter<'_>) -> sp_std::fmt::Result {
+		write!(f, "EcdsaSignature({:?})", &self.0[..])
+	}
+}
+
+type BalanceOf<T> = <<T as Trait>::Currency as Currency<<T as frame_system::Trait>::AccountId>>::Balance;
+
+pub trait Trait: frame_system::Trait {
+	type Event: From<Event<Self>> + Into<<Self as frame_system::Trait>::Event>;
+
+	/// The Currency for managing Evm account assets.
+	type Currency: Currency<Self::AccountId> + ReservableCurrency<Self::AccountId>;
+
+	/// Deposit for opening account, would be reserved until account closed.
+	type NewAccountDeposit: Get<BalanceOf<Self>>;
+
+	/// Mapping from address to account id.
+	type AddressMapping: AddressMapping<Self::AccountId>;
+
+	/// Handler to kill account in system.
+	type KillAccount: Happened<Self::AccountId>;
+
+	/// Weight information for the extrinsics in this module.
+	type WeightInfo: WeightInfo;
+}
+
+decl_event!(
+	pub enum Event<T> where
+		<T as frame_system::Trait>::AccountId,
+		EvmAddress = EvmAddress,
+	{
+		/// Mapping between Substrate accounts and EVM accounts
+		/// claim account. \[account_id, evm_address\]
+		ClaimAccount(AccountId, EvmAddress),
+	}
+);
+
+decl_error! {
+	/// Error for evm accounts module.
+	pub enum Error for Module<T: Trait> {
+		/// Bad signature
+		BadSignature,
+		/// Bad address
+		BadAddress,
+		/// Account ref count is not zero
+		NonZeroRefCount,
+		/// Account still has active reserved
+		StillHasActiveReserved,
+	}
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as EvmAccounts {
+		pub Accounts get(fn accounts): map hasher(twox_64_concat) EvmAddress => T::AccountId;
+		pub EvmAddresses get(fn evm_addresses): map hasher(twox_64_concat) T::AccountId => EvmAddress;
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		type Error = Error<T>;
+		fn deposit_event() = default;
+
+		#[weight = T::WeightInfo::claim_account()]
+		pub fn claim_account(origin, eth_signature: EcdsaSignature, eth_address: EvmAddress) {
+			with_transaction_result(|| {
+				let who = ensure_signed(origin)?;
+
+				// recover evm address from signature
+				let address = Self::eth_recover(&eth_signature, &who.using_encoded(to_ascii_hex), &[][..]).ok_or(Error::<T>::BadSignature)?;
+				ensure!(eth_address == address, Error::<T>::BadAddress);
```

Comment:
```suggestion
				ensure!(eth_address == address, Error::<T>::BadSignature);
```

or `InvalidSignature`